tests/testsuite/rustflags.rs

```diff
@@ -272,6 +272,47 @@ fn env_rustflags_build_script_with_target() {
         .run();
 }
 
+#[cargo_test]
+fn env_rustflags_build_script_with_target_doesnt_apply_to_host_kind() {
+    // RUSTFLAGS should *not* be passed to rustc for build scripts when --target is specified as the
+    // host triple even if target-applies-to-host-kind is enabled, to match legacy Cargo behavior.
+    // In this test if --cfg foo is passed the build will fail.
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.0.1"
+                build = "build.rs"
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .file(
+            "build.rs",
+            r#"
+                fn main() { }
+                #[cfg(foo)]
+                fn main() { }
+            "#,
+        )
+        .file(
+            ".cargo/config.toml",
+            r#"
+                target-applies-to-host = true
+            "#,
+        )
+        .build();
+
+    let host = rustc_host();
+    p.cargo("build --target")
+        .masquerade_as_nightly_cargo()
+        .arg(host)
+        .arg("-Ztarget-applies-to-host")
+        .env("RUSTFLAGS", "--cfg foo")
+        .run();
+}
```

Comment:
This at least doesn't match my mental model of Cargo to the point that one of the checks when calculating rustflags may need adjusting. At least in my own mental model of what's happening here target-specific rustflags have been configured, and it's been explicitly opted-in to that `target-applies-to-host`, meaning that the target-specific rustflags should be passed to the host, so `--cfg foo` should be passed to the build script.

Reply:
I read the other comment first, so left an explanation there. It basically comes down to whether `= true` should match the _current_ Cargo behavior (in which case I argue it should be renamed), or whether `= true` should also pick up cases like this (which the original draft of this PR did) and some third state (`= "mixed"`) should map to the current behavior.